Taking a dependency on a recent nuget version?

- [x] This issue is blocking <!-- Describe below what is blocked. -->
- [ ] This issue is causing unreasonable pain

<!-- Write your issue description below. -->

As part of making .NET SDK reproducible/deterministic, I am working on fixing up nuget packages to be reproducible/deterministic too. NuGet is getting a new API to make this possible: https://github.com/NuGet/Home/pull/14785

Arcade creates nuget packages in `src/Microsoft.DotNet.Build.Tasks.Packaging/src/NuGetPack.cs` via the `PackageBuilder` API.

How can I consume this very-recent version of NuGet.Client in arcade to get access to the new API?

Do you know how to do this locally, using the VMR?

I tried building the VMR with the nuget PR applied. Then I tried using it to build itself with this additional change:

```

diff --git a/src/arcade/src/Microsoft.DotNet.Build.Tasks.Packaging/src/NuGetPack.cs b/src/arcade/src/Microsoft.DotNet.Build.Tasks.Packaging/src/NuGetPack.cs
index b44012221b7..eb3dead0360 100644
--- a/src/arcade/src/Microsoft.DotNet.Build.Tasks.Packaging/src/NuGetPack.cs
+++ b/src/arcade/src/Microsoft.DotNet.Build.Tasks.Packaging/src/NuGetPack.cs
@@ -138,6 +138,12 @@ public bool Deterministic
             set;
         }

+        public string DeterministicTimestamp
+        {
+            get;
+            set;
+        }
+
         public override bool Execute()
         {
             if (Nuspecs == null || Nuspecs.Length == 0)
@@ -260,7 +266,10 @@ public void Pack(string nuspecPath, string nupkgPath, Manifest manifest, bool pa
             bool creatingSymbolsPackage = packSymbols && (Path.GetExtension(nupkgPath) == _symbolsPackageExtension);
             try
             {
-                PackageBuilder builder = new PackageBuilder(deterministic: Deterministic);
+                PackageBuilder builder = new PackageBuilder(deterministic: Deterministic)
+                {
+                    DeterministicTimestamp = DeterministicTimestamp,  // This is the new API
+                };

                 string baseDirectoryPath = (string.IsNullOrEmpty(BaseDirectory)) ? Path.GetDirectoryName(nuspecPath) : BaseDirectory;
                 builder.Populate(manifest.Metadata);
```

But this fails to build:

```
reproducible-dotnet/dotnet-stage-1/src/arcade/src/Microsoft.DotNet.Build.Tasks.Packaging/src/NuGetPack.cs(271,21): error CS0117: 'PackageBuilder' does not contain a definition for 'DeterministicTimestamp' [/home/omajid/devel/reproducible-dotnet/dotnet-stage-1/src/arcade/src/Microsoft.DotNet.Build.Tasks.Packaging/src/Microsoft.DotNet.Build.Tasks.Packaging.csproj::TargetFramework=net11.0]
```

--- a/src/Microsoft.DotNet.Build.Tasks.Packaging/src/NuGetPack.cs
+++ b/src/Microsoft.DotNet.Build.Tasks.Packaging/src/NuGetPack.cs
@@ -6,9 +6,10 @@ using NuGet;
 using NuGet.Versioning;
 using NuGet.Packaging;
 using System;
-using System.IO;
 using System.Collections.Generic;
+using System.IO;
 using System.Linq;
+using System.Reflection;
 using NuGet.Common;
 
 namespace Microsoft.DotNet.Build.Tasks.Packaging
@@ -138,6 +139,12 @@ namespace Microsoft.DotNet.Build.Tasks.Packaging
             set;
         }
 
+        public string DeterministicTimestamp
+        {
+            get;
+            set;
+        }
+
         public override bool Execute()
         {
             if (Nuspecs == null || Nuspecs.Length == 0)
@@ -261,6 +268,7 @@ namespace Microsoft.DotNet.Build.Tasks.Packaging
             try
             {
                 PackageBuilder builder = new PackageBuilder(deterministic: Deterministic);
+                SetDeterministicTimestamp(builder, DeterministicTimestamp);
 
                 string baseDirectoryPath = (string.IsNullOrEmpty(BaseDirectory)) ? Path.GetDirectoryName(nuspecPath) : BaseDirectory;
                 builder.Populate(manifest.Metadata);
@@ -325,6 +333,30 @@ namespace Microsoft.DotNet.Build.Tasks.Packaging
                 }
                 Log.LogError($"Error when creating nuget {packageType} package from {nuspecPath}. {e}");
             }
+        }
+
+        private void SetDeterministicTimestamp(PackageBuilder packageBuilder, string deterministicTimestamp)
+        {
+            // Use reflection to set the PackageBuilder.DeterministicTimestamp
+            // property. This property may or may not be available depending on
+            // the NuGet.Client version.
+
+            var type = packageBuilder.GetType();
+            var deterministicTimestampPropertyName = "DeterministicTimestamp";
+            var propertyInfo = type.GetProperty(deterministicTimestampPropertyName, BindingFlags.Public | BindingFlags.Instance);
+            if (propertyInfo == null)
+            {
+                Log.LogMessage(LogImportance.Low, $"{type.FullName} does not contain property {deterministicTimestampPropertyName}. Not setting {nameof(deterministicTimestamp)}.");
+                return;
+            }
+
+            if (!propertyInfo.CanWrite)
+            {
+                Log.LogWarning($"{type.FullName}.{deterministicTimestampPropertyName} is not writable.");
+                return;
+            }
+
+            propertyInfo.SetValue(packageBuilder, deterministicTimestamp);
         }
 
         private Manifest TransformManifestToPackedPackageManifest(Manifest manifest)

--- a/src/Microsoft.DotNet.Build.Tasks.Packaging/src/build/Packaging.targets
+++ b/src/Microsoft.DotNet.Build.Tasks.Packaging/src/build/Packaging.targets
@@ -1275,6 +1275,10 @@
 
     <Message Condition="'$(_SkipCreatePackage)' == 'true'" Text="Skipping package creation for $(NuSpecPath) because the following files do not exist: @(_missingFiles)" />
 
+    <PropertyGroup>
+      <DeterministicTimestamp Condition="'$(DeterministicTimestamp)' == '' AND '$(SOURCE_DATE_EPOCH)' != ''">$(SOURCE_DATE_EPOCH)</DeterministicTimestamp>
+    </PropertyGroup>
+
     <NuGetPack Nuspecs="$(NuSpecPath)"
                OutputDirectory="$(PackageOutputPath)"
                ExcludeEmptyDirectories="true"
@@ -1287,6 +1291,7 @@
                AdditionalSymbolPackageExcludes="@(AdditionalSymbolPackageExcludes)"
                PackedPackageNamePrefix="$(PackedPackageNamePrefix)"
                Deterministic="$(Deterministic)"
+               DeterministicTimestamp="$(DeterministicTimestamp)"
                Condition="'$(_SkipCreatePackage)' != 'true'"/>
 
     <!-- Create a marker that records the path to the generated package -->

